sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnnecessaryUsings.cs

```diff
@@ -134,6 +136,35 @@ public override void VisitAwaitExpression(AwaitExpressionSyntax node)
                 base.VisitAwaitExpression(node);
             }
 
+            /// <summary>
+            /// LINQ Query Syntax do not use symbols from the 'System.Linq' namespace directly, but the using directive is
+            /// still necessary to use the Query Syntax form.
+            /// </summary>
+            public override void VisitQueryExpression(QueryExpressionSyntax node)
+            {
+                if (!this.linqQueryVisited && TryGetSystemLinkNamespace(out var systemLinqNamespaceSymbol))
+                {
+                    this.necessaryNamespaces.Add(systemLinqNamespaceSymbol);
+                }
+                this.linqQueryVisited = true;
+                base.VisitQueryExpression(node);
+            }
+
+            private bool TryGetSystemLinkNamespace(out INamespaceSymbol systemLinqNamespace)
+            {
+                foreach (var usingDirective in this.usingDirectivesFromParent)
+                {
+                    if (this.context.SemanticModel.GetSymbolInfo(usingDirective.Name).Symbol is INamespaceSymbol namespaceSymbol &&
```

Comment:
It is needed as the `necessaryUsings` lists gather `INamespaceSymbol`. However I switched the conditions for efficiency, checking first the name based on the syntax node, and only retrieving it when necessary. (Note that this is also the reason I added a flag, in order to avoid looking for the symbol again if it was found in current context)

Reply:
I saw the flag and I liked it.